plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixPageSource.java

```diff
@@ -0,0 +1,136 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.phoenix5;
+
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.connector.ConnectorPageSource;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.trino.spi.block.RowBlock.fromFieldBlocks;
+import static java.util.Objects.requireNonNull;
+
+public class PhoenixPageSource
+        implements ConnectorPageSource
+{
+    private final ConnectorPageSource delegate;
+    private final List<ColumnAdaptation> columnAdaptations;
+
+    public PhoenixPageSource(ConnectorPageSource delegate, List<ColumnAdaptation> columnAdaptations)
+    {
+        this.delegate = delegate;
+        this.columnAdaptations = columnAdaptations;
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        Page page = delegate.getNextPage();
+        if (page == null || columnAdaptations.isEmpty()) {
+            return page;
+        }
+
+        return getColumnAdaptationsPage(page);
+    }
+
+    private Page getColumnAdaptationsPage(Page page)
+    {
+        Block[] blocks = new Block[columnAdaptations.size()];
+        for (int i = 0; i < columnAdaptations.size(); i++) {
+            blocks[i] = columnAdaptations.get(i).getBlock(page);
+        }
+
+        return new Page(page.getPositionCount(), blocks);
+    }
+
+    @Override
+    public long getMemoryUsage()
+    {
+        return delegate.getMemoryUsage();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        delegate.close();
+    }
+
+    public interface ColumnAdaptation
+    {
+        Block getBlock(Page sourcePage);
+
+        static ColumnAdaptation sourceColumn(int index)
+        {
+            return new SourceColumn(index);
+        }
+
+        static ColumnAdaptation mergedRowColumns(List<Integer> mergeRowIdSourceChannels)
+        {
+            return new MergedRowAdaptation(mergeRowIdSourceChannels);
+        }
+    }
+
+    private record MergedRowAdaptation(List<Integer> mergeRowIdSourceChannels)
```

Comment:
This runs into error prone check problem, I would just use a normal class here, it doesn't need to be a record

Reply:
Sure